### DISCUSSION TITLE
Proposal: Add `is_disconnected` property to `WebSocket`

### DISCUSSION THREAD

**Discussion by ruitcatarino** [encode/starlette#2762] [Ideas] [answered]
### Motivation
When handling a WebSocket connection, it’s common to have multiple coroutines interacting with the same connection. For example, one coroutine may handle incoming messages, while another handles error recovery or cleanup. In such scenarios, it's possible for multiple coroutines to attempt to close the WebSocket. This can lead to a `RuntimeError` if a coroutine attempts to close an already-disconnected WebSocket.

To prevent this, I propose adding an `is_disconnected` property to the `WebSocket` class. This property would allow developers to check the connection state before performing actions like close.

### Benefits
1. Improved Code Clarity
Developers can use `is_disconnected` to explicitly check the state of the WebSocket, avoiding reliance on exceptions for flow control. This results in cleaner, more maintainable code.
2. Concurrency Safety
In concurrent scenarios, checking `is_disconnected` before attempting an action reduces the risk of race conditions and redundant operations.
3. Minimal Overhead
The implementation is straightforward, as it wraps existing WebSocket states (`application_state` and `client_state`) into a single, intuitive property.

Consider this scenario (that already happened to me):
 - Two coroutines manage a WebSocket connection.
 - An error occurs, and both attempt to close the connection.
 - Without `is_disconnected`, one coroutine succeeds, while the other raises a `RuntimeError`.
By using `is_disconnected`, this scenario is easily handled:
```python
if not websocket.is_disconnected:
    await websocket.close()
```
### Implementation
Here’s a simple implementation of the `is_disconnected` property:
```python
@property
def is_disconnected(self) -> bool:
    return (
        self.application_state == WebSocketState.DISCONNECTED
        or self.client_state == WebSocketState.DISCONNECTED
    )
```

### Conclusion
Maybe this is too specific or maybe even based on a wrong principle or a problem with my app or thinking.
I’d love to hear your thoughts on whether this aligns with the framework’s goals and design.
I would gladly implement this if we want to move forward.

**Comment by adriangb** [comment]
I feel like this may not be the right answer. For example, we could add `ws.close(ignore_if_disconnected=True)` or something instead.

**Reply by Kludex** [reply]
The `ws.close()` is a noop already if disconnected... 🤔

**Reply by ruitcatarino** [reply]
Thank you for the time and feedback! 

While `ws.close()` is a no-op if already disconnected, it still raises a `RuntimeError`, which would need to be handled (or, in my case, ignored).

Both the `is_disconnected` property and the `ignore_if_disconnected=True` approach would let developers check the state explicitly, avoiding reliance on exceptions for flow control. It’s a small addition that could potentially improve code clarity, if you think it aligns with the library's goals.

**Reply by Kludex** [reply]
I think it only raises if you call close twice... I'll check later.

**Reply by Kludex** [reply]
Yeah, I'm right.

Maybe we can add a conditional on the `WebSocket.send()` to handle `self.application_state == DISCONNECTED`, and raise `WebSocketDisconnect` if the application is disconnected.

Code I used:

```py
from starlette.applications import Starlette
from starlette.routing import WebSocketRoute
from starlette.websockets import WebSocket


async def websocket(ws: WebSocket):
    await ws.accept()
    await ws.send_text("Hello, world!")
    await ws.close()
    # Call this to see the RuntimeError.
    # await ws.close()


app = Starlette(debug=True, routes=[WebSocketRoute("/ws", websocket)])
```

Can you share an MRE, or a small sample from your code that I can run, to see if this is really needed?

**Reply by ruitcatarino** [reply]
Hi there I can show a simple example where I do get the `RuntimeError` using `FastAPI`
```python
@app.websocket("/test")
async def test(websocket: WebSocket):
    await websocket.accept()
    await websocket.close()
    await websocket.close()
```
![image](https://github.com/user-attachments/assets/3d6ac501-70fe-4e8f-93de-06d747d1bda6)

**Reply by Kludex** [reply]
...It's the same code as the one I shared above.

Just don't call `close` twice...

**Reply by ruitcatarino** [reply]
I know I'm not supposed to call `close` twice... Like I explained in my initial question/proposal I don't want to, that's the entire purpose of this discussion.
```python
async def example_task_1(websocket: WebSocket):
    await asyncio.sleep(2) # Something happens
    await websocket.close()

async def example_task_2(websocket: WebSocket):
    await asyncio.sleep(2) # Something happens
    await websocket.close()

@app.websocket("/test")
async def test(websocket: WebSocket):
    await websocket.accept()
    await asyncio.gather(
        example_task_1(websocket),
        example_task_2(websocket),
    )
```
In this case one of the coroutine will raise an `RuntimeError`. What I want is a way to verify if the `websocket` is still open/connected before calling call `close`/`send_text`/`receive_text`/etc.. So something like this:
```python
async def example_task_1(websocket: WebSocket):
    await asyncio.sleep(2) # Something happens
    if not websocket.is_disconnected():
        await websocket.close()
```

**Reply by Kludex** [reply]
I need a real use case to help.

On your example, you can work something out to close the connection after one of the example tasks finish.

**Reply by ruitcatarino** [reply]
Here’s an example where `is_disconnected` is helpful. In this case, I want to ensure only one active WebSocket connection per endpoint. If a new client connects, the existing connection is closed gracefully.

```python
class SingleWebSocketManager:
    def __init__(self):
        self.active_websocket: WebSocket | None = None

    async def close_active(self):
        if self.active_websocket and not self.active_websocket.is_disconnected:
            await self.active_websocket.close()
        self.active_websocket = None

    async def handle_websocket(self, websocket: WebSocket):
        await self.close_active()
        await websocket.accept()
        self.active_websocket = websocket
        try:
            while not websocket.is_disconnected:
                data = await websocket.receive_text()
                await websocket.send_text(f"Echo: {data}")
        except WebSocketDisconnect:
            pass
        finally:
            self.active_websocket = None
```

Using `is_disconnected` here ensures the WebSocket is only closed if it’s still active, avoiding a `RuntimeError`. Without it, I’d have to rely on exception handling, making an unnecessary call to `close` that would call `send` that would eventually raise the exception just for me to `pass`.

I sense some hesitation about adding this. If you feel it’s not helpful or aligned with the framework’s goals, I completely understand and respect that. Thank you!

**Reply by adriangb** [reply]
I think it's reasonable to want this feature. I will note that if we had made `close()` idempotent someone would have complained that it didn't raise an error.

I'd be happy to accept a PR for `is_disconnected()`, although that may need a bit of design work: does it just check the internal attribute or actively check if the connection is alive. Personally I would prefer `close(error_if_already_closed=False)` similar to `pathlib.Path(...).mkdir(exists_ok=True)`.

**Reply by Kludex** [reply]
I think it makes more sense to raise WebSocketDisconnect instead of RuntimeError when the application state is disconnected.

I don't think the flag makes sense.

**Reply by ruitcatarino** [reply]
@adriangb I completely understand why `close()` can’t be idempotent, and I agree that raising an error is the right default behavior. I also think your proposed alternative `close(error_if_already_closed=False)` is a great addition and would provide a clean, explicit way to handle this scenario. I will open an issue/PR and implement this (later today probably).

@Kludex I see your point about raising `WebSocketDisconnect` instead of `RuntimeError` when the WebSocket is disconnected, and I’d be happy to implement that change as well. It would make the error more specific and informative, addressing part of the problem.
That said, I still believe it would be helpful to provide either a way to proactively check the WebSocket state (via an `is_disconnected` property) or a flag to ensure that `close()` can always be called gracefully (like @adriangb suggested).

**Reply by Kludex** [reply, marked as answer]
I believe the exception is the way to go.
